### ISSUE TITLE
Launch programs from brickman

### ISSUE THREAD

**Issue by WasabiFan** [ev3dev/ev3dev#129]
This is just an idea, so it might not line up with the general goals of ev3dev, but I'll just put it out there in case it's useful.

 I have been thinking that because this is an OS for the EV3 (a programmable robot controller), maybe we should make some sort of a system for holding/registering programs that are specifically for the EV3. Something like the 'bin' folder on most Linux systems, that would only contain programs that were EV3-specific. This could enable display manager integration, so one could navigate through a list of programs and run them like one would in the stock firmware, except the programs could be written in any language. Or we could set up some sort of a service to install programs on the EV3 from a host computer, without the user needing to have any knowledge of the Linux internals (no SSH sessions, no NFS shares, etc). Again, this might be more of an 'accessory' that would distract from some of the other features, but it seems like it could be pretty interesting.

I think this could help ev3dev feel more like an OS built for the EV3, instead of just another Linux distro with the ability to control the brick's devices. Either way, ev3dev works well functionality-wise as it currently stands; I think this would just be a higher level of brick-specific customization.

ADMIN EDIT FOR ISSUE TRACKING:
- [x] File Browser that can run user programs.
- [ ] App menu that runs programs (user or installed from package).

**Comment by pabr** [comment]
As far as I know there is no easy way for novice users to install software on ev3dev. ssh/scp/nfs over usb-ethernet/usb-cdc/usb-rndis/bnep/wifi does work and is nicely documented on the wiki, but this is a huge barrier to entry for non-technical users.

So, if application management functionality is added to brickman, please consider having it scan the Windows partition of the microSD card for software to install. Any kid with a smartphone knows how to download a file from the Internet and write it to a memory card.

As a quick-dirty-and-dangerous solution, the init scripts could automatically install any deb file found in /media/mmc_p1 or execute /media/mmc_p1/autorun.sh if it exists. 

I understand that ev3dev aims to demystify robot programming and it would be bad if users only use it to run binary blobs. But most EV3 owners are not going to bother testing ev3dev if they have to learn Linux and networking from scratch before they see any benefit.

**Comment by dlech** [comment]
Copying files over ssh is as easy as drag-and-drop with programs like WinSCP and FileZilla. I do agree that network setup could be easier though. In the last couple of kernel versions, we have enabled the modules required for ConnMan to do tethering. With tethering over USB or Bluetooth, there should be no configuration required on the host computer. It should "just" work.

Some other options are ...

1. Bluetooth file transfer (OBEX).
2. USB Mass Storage Gadget. The catch on this one is that the partition used for the mass storage device cannot be mounted on the EV3 and the host computer at the same time or else file corruption *will* happen.

And yes, we have taken the Arch Linux approach/philosophy so far in that we expect users to learn a bit about administering Linux in order to be able to use ev3dev. From @rhempel's [original vision](http://hempeldesigngroup.com/embedded/2013/09/25/making-hacking-accessible-lego-mindstorms-ev3/):

> What I’ll be releasing allows the Level 2 Daily Computer Users to finally get a handle on real programming tools for robots. They might need some help from a Level 3 [Computer Programmer] friend to set up the microSD card the first time, but after that it’s a piece of cake.


----

There is also the point that has been brought up before of "what about the people that want to just run the awesome program that someone else wrote for ev3dev, but not actually use ev3dev for programming themselves". Perhaps this is more of what you are getting at. In this case, I think integrating Bluetooth OBEX and USB Mass Storage gadget into brickman would be the way to go. It doesn't get much easier than that.

**Comment by susi** [comment]
A possibility to launch programs from brickman would indeed be a dream come true for me. What's the current status? Is there anything I could help with regarding getting this done? I'm most interested in being able to launch programs from the brickman menu, and I think that a .desktop file sounds like a good standard approach.

**Comment by dlech** [comment]
I haven't started on the implementation yet - although I am currently tinkering with how do we hand off console ownership to the new program. I am leaning towards option 2. from <https://github.com/ev3dev/ev3dev/issues/129#issuecomment-58698448>. I still have some more research to do before I can make a decision though.

If you are interested in hacking on brickman, see <https://github.com/ev3dev/brickman>.

Also see the discussion in #232. If you just want to design how it should look and how it should work from the user's point of view, that would be helpful too.

**Comment by susi** [comment]
Great, I'll have a look. I can draw you some screens for selecting programs to run. I'm thinking that brickman would be looking for .desktop files is a specific directory, like /var/lib/brickman/programs (I can also draw display where you are browsing the whole filesystem).

I was thinking that since coding the display seems to be a bit annoying and the same for the buttons, would it be possible to abstract this code into a "libbrick", to make it easier for application developers to use the buttons and console screen? The hard part here is of course language support...

So, as an alternative, brickman could present a D-BUS API that programs can use to draw things on the screen, and also register for button press events. This would though mean that they would run on the VT of brickman, which would need to pause it's drawing functions, but it would be relatively easy for application programmers to draw stuff on the screen and react to button presses without needing to worry about low-level stuff like managing the display and key events.

What do you think?

**Comment by dlech** [comment]
> I can also draw display where you are browsing the whole filesystem

I don't think that browsing the whole file system on the brick screen would be that useful.

> would it be possible to abstract this code into a "libbrick", to make it easier for application developers to use the buttons and console screen? The hard part here is of course language support...

This is already in progress. It is called [ev3devKit](https://github.com/ev3dev/ev3devKit). It uses GLib and GObjects, so it will be usable in just about any language through GObject Introspection. See the readme [here](https://github.com/dlech/ev3dev-lang-glib) for some links. There is also a lower level graphics library [GRX](https://github.com/ev3dev/grx) that ev3devKit is built on, but it could be used by itself. GRX actually has an input library (it is in the 'mouse` subdirectory, but it includes keyboard as well). However, it does not support separate key up and key down events, just key pressed. Right now though, I am using ncurses for keyboard (button) input in ev3devKit instead of GRX.

> So, as an alternative, brickman could present a D-BUS API that programs can use...

This has occurred to me as well. I like the idea of using D-Bus because it is super easy to implement. However, I also would like remote control programs (programs that run on another device, not the EV3) to be able to use the same API, so something like [JSON-RPC](http://json-rpc.org/) might make more sense.

And actually, we could do it both ways. i.e. the .desktop file has an entry that indicates which type the program is.

But, for starters, here is what I am thinking. We just start with the first option. No remote controlling / D-Bus yet. When running a program, brickman will take care of switching consoles. The user program will be responsible for putting the console in graphics mode if needed (the GRX library takes care of this). Brickman will monitor the process and switch the console back when the user program ends. Or, if the user presses and holds the back button for 5 seconds, brickman will send SIGINT to the users program.

Also, it would be nice if there was a command line tool for running user programs from a remote terminal, it would work the same way, except that it would redirect stdin/stdout/stderr to the remote console for debugging.

**Comment by dlech** [comment]
It's [official], you can now launch programs from brickman v0.4.0. I changed my mind about the file browser. This seemed to be the easiest way to get something implemented. Please report any problems with the implementation by opening *new* issues - don't report them in this issue.

I still like the idea of of having a separate "Apps" menu that is populated by the .desktop files that we have been discussing, so I am leaving this issue open for further design discussion about that.

[official]: http://www.ev3dev.org/news/2015/01/25/Package-Release/

**Comment by SciStudent** [comment]
I am trying to run a python-ev3 file with brickman. So I made a shell file including the line "python myfile.py". This file runs fine in ssh as root. Running the file as normal user requires "sudo" and password. Basically python-ev3 uses several system level resources that need root level access. 

Now since brickman prevents any root level program from running, there seems to be no way to run this python-ev3 file. When I made the shell file as root, I got stopped at brickman for "running root program". When I changed the file owner to a normal user, file was stopped for "trying to access system folder". 

Is there any suggestion for this situation? Have anyone using python-ev3 got the same problem?

**Comment by dlech** [comment]
> Basically python-ev3 uses several system level resources that need root level access.

It would be nice to know what these are. I imagine one is accessing I2C directly.

> Is there any suggestion for this situation?

Make sure your user is in the `sudo` group and then use `visudo` to change so that you can [use sudo without a password](http://jeromejaglale.com/doc/unix/ubuntu_sudo_without_password).

Then write a shell script in your (non-root) user directory that uses sudo to call your program.

    #!/bin/bash

    sudo python program.py

**Comment by SciStudent** [comment]
Thanks! It works! I just followed the instruction in the link and ran "sudo visudo" and added "my_user_name ALL=(ALL) NOPASSWD=ALL" at the end of the file. It worked! On both ssh and brickman. Previously I tried to use NOPASSWD= destination_file as someone recommended but that did not work.

As for what system call was used by Python-ev3, in my case, it was reset for MediumMotor. I was only testing the sample program there. My actual program will use USB serial to communicate with Raspberry Pi. So the USB port will be occupied and I need to start the program from brickman.
